index.js

```diff
@@ -89,3 +107,25 @@ function isEmail(email) {
     const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
     return re.test(String(email).toLowerCase())
 }
+
+function isAnonymousIdentifier(distinct_id) {
+    const re = /^[\w]{14}-[\w]{14}-[\w]{8}-[\w]{6}-[\w]{14}$/g
+    return re.test(String(distinct_id))
+}
+
+function getEmailFromEvent(event) {
+    if (isEmail(event.distinct_id)) {
+        return event.distinct_id
+    } 
+
+    const getEmailFromKey = (key) => {
+        const object = event[key]
+        if (!object) return false
+        if (!Object.keys(object).includes('email')) return false
+
+        const email = object['email']
+        return isEmail(email) && email
+    }
+    
+    return getEmailFromKey('$set') || getEmailFromKey('$set_once') || getEmailFromKey('properties')
+}
```

Comment:
I originally got this snippet from a different plug-in, but it wasn't checking for `$set_once`. In order to avoid repetition, a single function to check if the key is there and if it is an email.